### ISSUE TITLE
[bug] options of dependencies affect consumer package_id()

### ISSUE THREAD

**Issue by barmaglot27** [conan-io/conan#6987]
Related to #6840

All our packages follow semantic versioning and use `minor_mode()` for compatibility, so I was expecting package_id calculation to include: settings, its **own options** and name/major.minor of dependencies. But once I introduced an option with default value in our company's deepest dependency it broke consumer's package_id().

The problem seem to be related to `OptionsValues.sha()` method:
https://github.com/conan-io/conan/blob/bd5920a40d9707001b31966adb0dfa7d6af00eea/conans/model/options.py#L284-L290

it is invoked from `ConanInfo.package_id()`:
https://github.com/conan-io/conan/blob/bd5920a40d9707001b31966adb0dfa7d6af00eea/conans/model/info.py#L528-L543

### Steps to reproduce

1. Dependency tree:
    binary/1.0.0+0 -> libB/1.0.0+0 -> libA/1.0.0+0
    compatibility is set to minor version for all packages.
    computed hash of libB/1.0.0+0 is 1111 (from settings, libB options, libA/1.0.Z)

2. In next patch version of libA/1.0.1 introduce an option with default value variant=default and build new minor version:
    pkgA/1.0.1+0

3. Add libA/1.0.1+0 as explicit dependency of binary/1.0.1+0 to override and achieve this:
    binary/1.0.1+0 -> libA/1.0.1+0
    binary/1.0.1+0 -> libB/1.0.0+0 -> libA/1.0.**1**+0
    computed hash of libB/1.0.0+0 suddenly becomes **2222**
    According to docs it should be same (from settings, libB options, libA/1.0.Z), but I poked with debugger and clearly see that inside package_id() method of libB recipe `self.info.options.sha` returns different value for case 1. and 3. 

### Environment Details
  * Operating System+version: RHEL 7.2
  * Compiler+version: g++ 4.8.3
  * Conan version: 1.24.1
  * Python version: 3.7

**Comment by memsharded** [comment]
Hi @barmaglot27 

The ``minor_mode()`` or any ``package_id_mode`` is designed and defined to only affect what part of the dependencies ``package_reference`` affects the current computation of the package ID. That is, the modes are related to the ``self.info.requires`` attribute, but not to the ``self.info.options`` or ``self.info.settings``.

I agree this is unexpected and maybe not enough clearly documented. And I agree it would be better to get rid of this effect. Unfortunately, this cannot be done without massively breaking many users packages. so this will need to wait until Conan 2.0 to be properly defaulted.

Meanwhile, you could use this in your recipe:

```python
def package_id(self):
    ...
   self.info.options.clear_indirect()
```

**Comment by barmaglot27** [comment]
Thanks, I will add it to our conanfiles. Will have to re upload all dependencies though.

**Comment by barmaglot27** [comment]
@memsharded the suggested approach does not work! Simply because `self.info.options.clear_indirect()` does nothing to keys of the array:
https://github.com/conan-io/conan/blob/bd5920a40d9707001b31966adb0dfa7d6af00eea/conans/model/options.py#L251-L253

So, if any dependency adds option - it will still affect consumer's package id. The only solution would be some API that completely clears `OptionsValues._reqs_options`. I could not find such API, but following seems to fit the bill:
https://github.com/conan-io/conan/blob/bd5920a40d9707001b31966adb0dfa7d6af00eea/conans/model/options.py#L255-L256 

What do you think of:
```python
def package_id(self):
    ...
   self.info.options.filter_used([])
```

It seems to do what I want - at the time of `ConanInfo.package_id()` call there is another filter_used() call with list of requirements, which will be noop, and then during `OptionsValues.sha` only digest of `OptionsValues._package_values.sha` is taken into account, all `OptionsValues._reqs_options` are not present.

**Comment by memsharded** [comment]
You are totally right! I didn't look at the implementation. What you are pointing:

```python
def package_id(self):
    ...
   self.info.options.filter_used([])
```

Will do what you want. Great job.

**Comment by memsharded** [comment]
This has been solved in 2.0. Options of dependencies no longer affect consumer ``package_id()`` by default. So closing this as solved in 2.0.
